### PR TITLE
Custom particle generators for particle spawners

### PR DESCRIPTION
**Pull request by raymoo**
This PR adds a DSL for particle spawners to allow more flexibility in particle generation. It allows custom generation of position, velocity, acceleration, exptime, and size, and allows the values to be related to each other by having them use the same random generators. It also gives time as an input, so patterns can change over time.

Example video: https://a.safe.moe/KVBCU.webm
Code for example:
```
local expr = expression

local angle = expr.random(0, 2 * math.pi)
local circle = { x = expr.sin(angle), y = expr.random(0, 0.1), z = expr.cos(angle) }
local upspeed = expr.min(5, 5 * expr.time())
local velocity = { x = 0, y = upspeed, z = 0 }
local gravity = { x = 0, y = -10, z = 0 }

expr.trace({ expr.time() * circle, velocity, 2 * expr.time() })

minetest.override_item("default:stick", {
        on_use = function(stack, user)
                local pos = user:getpos()
                minetest.add_particlespawner({
                                amount = 1000 * 5,
                                time = 5,
                                minpos = pos,
                                maxpos = pos,
                                minacc = gravity,
                                maxacc = gravity,
                                pos_expression = expr.time() * circle,
                                vel_expression = velocity,
                                size_expression = 2 * expr.time(),
                                minexptime=0.5,
                                maxexptime=0.5,
                                texture = "default_lava.png",
                                glow = 15,
                                collisiondetection = true,
                })
        end,
})
```

### CODE REVIEW

**Comment by raymoo** [comment]
I think the build is failing because particles.cpp isn't being built before minetestserver is linked. How should I fix that?

**Comment by sofar** [comment]
```elseif self.gen_type == "constant_scalar" then```

Maybe make this a simple lookup table?

```
local trace_fn = trace_tbl[self.gen_type]
if trace_fn then
     ...
```

**Comment by raymoo** [comment]
@sofar It may or may not improve performance, but I think it may harm readability by increasing indirection in the code.

**Comment by raymoo** [comment]
I added documentation so this is no longer WIP. I still have not fixed the build issue with minetest server, though.

**Comment by orwell96** [comment]
Do the vectors you add up to the original position take the yaw of the object into account when the particle spawner is attached?
I once implemented the support for this, and it should be kept consistent.

**Comment by raymoo** [comment]
@orwell96 Yes, the yaw transformation is applied after the "shaders". I am getting new build problems after fixing the old ones, though.

**Comment by raymoo** [comment]
Also it's annoying because it looks like the build failures only happen in clang, and I test on gcc.

**Comment by raymoo** [comment]
Ok, looks like CI is making working builds now.

To put this feature in context, my original use case was creating particle spawners that made particles in a cone, for a magic mod. However, I felt that supporting just cones would be too specialized, so made this instead.

#4501 is where I started asking about making particle spawners more flexible.

**Comment by raymoo** [comment]
The linter is telling me to mix tabs and spaces so you can just ignore it.

**Comment by sofar** [comment]
Actually the linter is (annoyingly) correct. Your indent should be 7 spaces before `<<`. However, I don't see it as a thing that should prevent the patch from getting through.

I looked throught the code fairly cautiously and I certainly approve the concept (:+1:) and API (:+1:), but I'd like one of the people who know C++ more than my potato level to skim it over as well ;)

**Comment by raymoo** [comment]
It looked like to me the linter was suggesting a tab and 3 spaces. In either case, I thought our codebase only uses tabs for indentation?

**Comment by sofar** [comment]
```
-		<< serializeString(exptime) << serializeString(size);
+	       << serializeString(exptime) << serializeString(size);
```

It's suggesting a tab and 7 spaces - this is directly pasted from the linter output on the webpage. Do you have tabs set to something else than 8 wide, perhaps?

**Comment by raymoo** [comment]
@sofar I was doing it from the same place. I didn't remember how many spaces it was, but I thought you were saying that the linter was saying to use 7 spaces (and no tabs), so I assumed the tab I had seen was accounting for four spaces.

**Comment by rubenwardy** [comment]
I actually quite like this dsl. Random generator is the wrong name however, I feel. Maybe delayed expression or expression or similar

**Comment by raymoo** [comment]
@rubenwardy I was unsure about the naming. I am calling it expression now.

File: src/particles.cpp [hunk 1]
```diff
@@ -31,15 +32,23 @@ with this program; if not, write to the Free Software Foundation, Inc.,
...
 
+#include <cmath>
+#include <stack>
+
 /*
 	Utility
 */
 
-v3f random_v3f(v3f min, v3f max)
+static inline f32 random_f32(f32 min, f32 max)
```
Comment (red-001) [inline src/particles.cpp]: wouldn't it be better to move this to `util/numeric`?

**Comment by raymoo** [comment]
* Even with this, having some kind of bulk add particle could be useful to implement particle generation patterns that aren't possible with this VM
* Okay
* True, and there are some things that I know could have better performance. For example, I represent scalar values as a vector where all three values are the same, so that I don't need a type tag for the values in the VM. Also there are no branching primitives in the expression language, so any conditionals will need to run all branches.
* This PR only affects the generation of particles, so as long as individual particle positions, velocities, accelerations, sizes, expiration times, etc. can be specified on new particles, it should be compatible.

**Comment by raymoo** [comment]
Any news on whether core devs think this is a desirable feature / implementation approach?

After thinking a bit, there are some things I would change about the current particle language / VM:
- Add branching (ternary operator).
  - (Maybe) Add multi-way branching. For example, if the condition is <=0, use expression 0, if it's <=1, use expression 1, if it's <=2, use expression 2. Can be more efficient than two-way branching if you want to have a particle be generated randomly as one of a few types of particle, or if you want to define the behavior of the particle in a piecewise way.
- Examine the list of built-in math functions to see if more would be useful.
- (Maybe) Perform operations on scalars more efficiently. Currently scalar values are represented as v3 and operations cost the same as on vectors.

**Comment by paramat** [comment]
Sorry to write this but, i prefer that Lua is used with individual particles to do complex behaviour, and that any really useful new behaviours are added to particle spawner code in the standard way.
What this PR allows is limited and this will be the case even with more possibilities added. Modders have precise needs and will very often end up wanting a behaviour this doesn't do. This seems the wrong approach, too hardcoded and overcomplex. Also, improving particles is high priority and should be done first. If this makes that more difficult to do that's also bad.
:-1:

**Comment by raymoo** [comment]
If Lua is the way to extend particles, the following should be implemented:
- Server provided client side mods (so that particle spawning can be handled by client-side Lua and not suffer from network lag or put load on the server)
- Batch version of `add_particle` that takes a list of particles and adds them all at once (to avoid the overhead of one API call per new particle)
  - Batch version where you can specify some parameters that are uniform over the batch-added particles would be nice and could be more efficient

I agree that creating particles from Lua would be more flexible, and you would also have access to more information about the world than this DSL currently provides. Half of my reason for writing this PR was for fun, so I won't mind too much if it's rejected in favor of CSM-based particle generation. The other half is that I didn't want to wait for server-provided CSM, which I may or may not have sufficient time to contribute to.

If server-provided CSM and particle batching gets in, then this API could be implemented as a mod using those features and a globalstep callback as the underlying implementation. The performance might be lower, but I don't think that it would be a significant cost at the volumes of particles that mods output.

**Comment by raymoo** [comment]
I don't think this would make particles improvement significantly more difficult, though. All it currently affects is particle generation.

**Comment by SmallJoker** [comment]
It's an impressive amount of work you put into this PR. It shall not be lost; maybe particle generators should be exposed to CSMs to modify them there instead of relying on a new interpreter solely designed for particles.

Feel free to re-open any time if the complexity could be reduced in some way.

**Comment by raymoo** [comment]
I don't think the complexity of this version of this solution to custom particle spawners can be significantly reduced. If SSCSMs get in and they can spawn particles in batch, that would be good enough for me (individual particles is not good enough because it could be slow to make one engine call per new particle). SSCSMs would even be more flexible since you can use any Lua computation and use any inputs from the engine that are exposed through the modding API.

**Pull request closed without merging**